models/dpr_test/dpr_test.c

```diff
@@ -0,0 +1,238 @@
+  if (status != 0)
+    goto bad_alloc_chrdev_region;
+  sprintf(deviceMinor, "%d", MAJOR(dev_num));
+  strcat(device_name, deviceMinor);
+  pr_info("%s\n", device_name);
+  cl = class_create(THIS_MODULE, device_name);
+  if (cl == NULL)
+    goto bad_class_create;
+  cdev_init(&fe_dpr_test_devp->cdev, &fe_dpr_test_fops);
+  status = cdev_add(&fe_dpr_test_devp->cdev, dev_num, 1);
+  if (status != 0)
+    goto bad_cdev_add;
+ device_obj = device_create_with_groups(cl, NULL, dev_num, NULL, dpr_test_groups, device_name);
+  if (device_obj == NULL)
+    goto bad_device_create;
+  dev_set_drvdata(device_obj, fe_dpr_test_devp);
+  pr_info("dpr_test exit\n");
+  return 0;
+bad_device_create:
+bad_cdev_add:
+  cdev_del(&fe_dpr_test_devp->cdev);
+bad_class_create:
+  class_destroy(cl);
+bad_alloc_chrdev_region:
+  unregister_chrdev_region(dev_num, 1);
+bad_mem_alloc:
+bad_exit_return:
+  pr_info("dpr_test_probe bad exit\n");
+  return ret_val;
+}
+
+
+static int dpr_test_open(struct inode *inode, struct file *file) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+
+static int dpr_test_release(struct inode *inode, struct file *file) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+
+static ssize_t dpr_test_read(struct file *file, char *buffer, size_t len, loff_t *offset) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+
+static ssize_t dpr_test_write(struct file *file, const char *buffer, size_t len, loff_t *offset) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+static ssize_t name_read(struct device *dev, struct device_attribute *attr, char *buf) {
+  fe_dpr_test_dev_t * devp = (fe_dpr_test_dev_t *)dev_get_drvdata(dev);
+  sprintf(buf, "%s\n", devp->name);
+  return strlen(buf);
+}
+
+static ssize_t dpr_data_read(struct device *dev, struct device_attribute *attr, char *buf) {
+  fe_dpr_test_dev_t * devp = (fe_dpr_test_dev_t *)dev_get_drvdata(dev);
+  fp_to_string(buf, devp->dpr_data, 0, false, 32);
+  strcat2(buf,"\n");
+  return strlen(buf);
+}
+
+static ssize_t dpr_data_write(struct device *dev, struct device_attribute *attr, const char *buf, size_t count) {
+
+  int i,c;
+	uint32_t num;
+  char substring[max_char];
+  int substring_count = 0;
+  int num_counter = 0;
+  
+  fe_dpr_test_dev_t *devp = (fe_dpr_test_dev_t *)dev_get_drvdata(dev);
+
+  // Check to see if the function was called with an empty string
+  if (count < 2)
+    return count;
+    
+  for (i = 0; i < count; i++)
+  {
+    //If its not a space or a comma, add the digit to the substring
+    if ((buf[i] != ',') && (buf[i] != ' ') && (buf[i] != '\0') && (buf[i] != '\r') && (buf[i] != '\n'))
+    {
+      substring[substring_count] = buf[i];
+      substring_count++;
+    }
+    else 
+    {
+      substring[substring_count] = '\0';
+      num = set_fixed_num(substring,8,false);
+      iowrite32((num & 0x0000FFFF) + (((uint32_t)num_counter)<<16), (u32 *)devp->regs + 0);
```

Comment:
Having some comments about why this bitmask and the bitshift are being used would be good, as I often find low-level bit operations to be somewhat opaque in terms of intention.